pennylane/fourier/reconstruct.py

```diff
@@ -0,0 +1,295 @@
+    shifts_neg = -shifts_pos[::-1]
+    fun_at_zero = fun(qml.math.array(0.0)) if fun_at_zero is None else fun_at_zero
+    evals = list(map(fun, shifts_neg)) + [fun_at_zero] + list(map(fun, shifts_pos))
+    shifts = qml.math.concatenate([shifts_neg, [0.0], shifts_pos])
+
+    def _reconstruction(x):
+        """Univariate reconstruction based on equidistant shifts and Dirichlet kernels."""
+        return qml.math.tensordot(
+            qml.math.sinc(a * (x - shifts)) / qml.math.sinc(b * (x - shifts)),
+            evals,
+            axes=[[0], [0]],
+        )
+
+    return _reconstruction
+
+
+_warn_text_fun_at_zero_ignored = (
+    "The provided value of the function at zero will be ignored due to the "
+    "provided shift values. This may lead to additional evaluations of the "
+    "function to be reconstructed."
+)
+
+
+def _reconstruct_gen(fun, spectrum, shifts=None, fun_at_zero=None):
+    r"""Reconstruct a univariate (real-valued) Fourier series with given spectrum.
+    Args:
+        fun (callable): Fourier series to reconstruct with signature ``float -> float``
+        spectrum (Sequence): Frequency spectrum of the Fourier series; non-positive
+            frequencies are ignored
+        shifts (list): Shift angles at which to evaluate ``fun`` for the reconstruction
+            Chosen equidistantly within the interval :math:`[0, 2\pi/f_\text{min}]` if ``shifts=None``
+            where :math:`f_\text{min}` is the smallest frequency in ``spectrum``.
+        fun_at_zero (float): Value of ``fun`` at zero. If :math:`0` is among the ``shifts``
+            and ``fun_at_zero`` is provided, one evaluation of ``fun`` is saved.
+
+    Returns:
+        callable: Reconstructed Fourier series with :math:`R` frequencies in ``spectrum``,
+        as ``qml.numpy`` based function and coinciding with ``fun`` on :math:`2R+1` points.
+    """
+    # pylint: disable=unused-argument
+
+    have_fun_at_zero = fun_at_zero is not None
+    have_shifts = shifts is not None
+    # For an empty/trivial spectrum, the function simply is constant
+    if spectrum in ([], [0.0]):
+        if have_shifts:
+            fun_value = fun(shifts[0])
+            if have_fun_at_zero:
+                warnings.warn(_warn_text_fun_at_zero_ignored)
+        else:
+            fun_value = fun_at_zero if have_fun_at_zero else fun(0.0)
+
+        def constant_fn(x):
+            """Univariate reconstruction of a constant Fourier series."""
+            return fun_value
+
+        return constant_fn
+
+    spectrum = np.array([f for f in spectrum if f > 0.0])
+    f_max = max(spectrum)
+
+    # If no shifts are provided, choose equidistant ones
+    if not have_shifts:
+        R = len(spectrum)
+        shifts = np.arange(-R, R + 1) * 2 * np.pi / (f_max * (2 * R + 1)) * R
+        zero_idx = R
+        need_fun_at_zero = True
+    elif have_fun_at_zero:
+        zero_idx = np.where(np.isclose(shifts, 0.0))[0]
```

Comment:
How come `isclose` is required here?

Reply:
If the shifts are computed automatically somewhere and not cleanly set to `0` but to `1e-9`, we wouldn't want another QNode execution to be triggered, I thought :) One could discuss the optimal setting for `atol` for this `isclose`...

Reply:
Main reason I asked is that I realize it will block an attempt to JIT :( Since tensor values are not known at compile time

Reply:
Oh actually, I think it should be fine... since there is no Python branching dependent on the tensor value.

Reply:
I'm not entirely sure..But in any case where we want to handle user-input shifts that in addition are not necessarily sorted with 0 as first entry if it is in there (and even then we would get branching), we will have to perform some check of this kind, right?

Reply:
I recently added [`qml.math.is_abstract`](https://github.com/PennyLaneAI/pennylane/blob/c143264155d56cf78728472d961746d8c2bef482/pennylane/math/utils.py#L203), which is a bit of a 'stopgap' solution --- check if the tensor has a value. If it doesn't, we are likely in a JIT, and avoid any branching/validation based on tensor value.

Reply:
I say 'stopgap' because it is not ideal to _avoid_ branching logic in a JIT. Ideally, we would find a way to code it up so that it works universally 🤔

Reply:
Maybe I'm a bit slow here but how would we avoid branching, conditional on what kind of object is passed to the function? :thinking:

Reply:
ah, the branching is only not allowed for 'dynamic' tensor values :)

E.g.,

```python
@jax.jit
def cost(x, y):
    if y > 0.5:
        x = 2 * x * y
    return jax.numpy.cos(x)
```

```pycon
>>> cost(0.2, 0.6)
ConcretizationTypeError: Abstract tracer value encountered where concrete value is expected: Traced<ShapedArray(bool[], weak_type=True)>with<DynamicJaxprTrace(level=0/1)>
The problem arose with the `bool` function.
While tracing the function cost at <ipython-input-70-c16c294dc5fd>:1 for jit, this concrete value was not available in Python because it depends on the value of the argument 'y'.
```

But we can set `y` as 'static':

```python
@functools.partial(jax.jit, static_argnums=[1])
def cost(x, y):
    if y > 0.5:
        x = 2 * x * y
    return jax.numpy.cos(x)
```

in which case it will work:

```pycon
>>> cost(0.2, 0.6)
DeviceArray(0.971338, dtype=float32, weak_type=True)
```

`static_argnum` is basically a 'contract'; we are telling JAX that `y` is not a dynamic/trainable parameter, and doesn't need to be traced with an abstract tensor.

Essentially, we can have Python branches (`if`, `for`) that depend on tensor values _only if those tensor values are 'static'_.

Reply:
Ah, I see, thanks for the explanation! :)
I'm not sure though whether the shifts can be considered static? :thinking: Maybe for now they are.

Reply:
I'm thinking now, that maybe JITting is complicated anyways because there is a callable in the input arguments of `reconstruct`?

Reply:
I think it is fine - I had a look at the code again, and the only branching occurs with `have_shifts` and `have_f0`, which are fine to branch on --- these are just Python booleans, and aren't dependent on the tensor value.

The only problematic branch I can see is 

```python
cond = qml.math.linalg.cond(C)
if cond > 1e8:
```
since `cond` _is_ dependent on the value of a tensor!

Reply:
> I'm thinking now, that maybe JITting is complicated anyways because there is a callable in the input arguments of reconstruct?

Ah, yes this might definitely have an effect!